lib/cn/cn_tree.c

```diff
@@ -1965,20 +1961,262 @@ static merr_t
+    }
+
+    minnode = route_map_lookup(map, minkey, minklen);
+    assert(minnode);
+
+    if (route_node_keycmp(minnode, maxkey, maxklen) >= 0)
+        return route_node_tnode(minnode);
+
+    return NULL;
+}
+
+/* TODO: This function shares a fair bit of code with cn_comp_spill().
+ */
+static merr_t
+cn_zspill(struct cn_compaction_work *w, bool *zspilled)
+{
+    struct kvset_list_entry *first, *le = w->cw_mark;
+    struct cn_tree *tree = w->cw_tree;
+    struct cn_tree_node *n, *znode;;
+    struct subspill *ss_arr;
+    struct route_node *rtn = NULL;
+    atomic_uint *spillingp = NULL;
+    atomic_uint *zspillingp = NULL;
+    void *lock;
+    uint i;
+    merr_t err = 0;
+
+    *zspilled = false;
+    znode = NULL;
+
+    rmlock_rlock(&tree->ct_lock, &lock);
+
+    /* Confirm that the operation can still be a zspill.
+     */
+    for (i = 0, first = le; i < w->cw_kvset_cnt; ++i) {
+        first = le;
+
+        assert(kvset_get_workid(le->le_kvset) != 0);
+        n = cn_kvset_can_zspill(le->le_kvset, tree->ct_route_map);
+        if (!n)
+            break;
+
+        if (!znode)
+            znode = n;
+        else if (n->tn_nodeid != znode->tn_nodeid)
+            break;
+
+        le = list_prev_entry(le, le_link);
+    }
+
+    if (i < w->cw_kvset_cnt) {
+        rmlock_runlock(lock);
+        return 0; /* Input kvsets no longer map to a single node. Cannot zspill, bail out */
+    }
+
+    /* Cannot zspill if znode is splitting or joining.
+     */
+    mutex_lock(&tree->ct_ss_lock);
+    if (znode->tn_ss_splitting || znode->tn_ss_joining) {
+        mutex_unlock(&tree->ct_ss_lock);
+        rmlock_runlock(lock);
+        return 0;
+    }
+
+    /* Since znode is neither splitting nor joining, the operation can continue as a zspill. Mark
+     * the target child node of the zspill as busy so it may be protected from splits and joins.
+     */
+    zspillingp = &znode->tn_ss_spilling;
+    atomic_inc_acq(zspillingp);
+    mutex_unlock(&tree->ct_ss_lock);
+    rmlock_runlock(lock);
+
+    ss_arr = calloc(CN_FANOUT_MAX, sizeof(*ss_arr));
+    if (!ss_arr)
+        return merr(ENOMEM);
+
+    i = 0;
+    while (1) {
+        struct route_node *rtnext;
+        struct subspill *ss;
+
+        rmlock_rlock(&tree->ct_lock, &lock);
+
+        rtnext = rtn ? route_node_next(rtn) : route_map_first_node(tree->ct_route_map);
+        if (!rtnext)
+            break;
+
+        n = route_node_tnode(rtnext);
+        if (!n->tn_route_node)
+            abort();
+
+        mutex_lock(&tree->ct_ss_lock);
+        if (n->tn_ss_splitting || n->tn_ss_joining < 0) {
+            ss = list_last_entry_or_null(&n->tn_ss_list, typeof(*ss), ss_link);
+
+
+            /* It was established upfront that the znode is not undergoing a spill or a join.
+             */
+            if (n->tn_nodeid != znode->tn_nodeid && (!ss || w->cw_sgen > ss->ss_sgen)) {
+                char *wmesg = n->tn_ss_splitting ? "spltwait" : "joinwait";
```

Comment:
const

Reply:
Move setting of wmesg to after dropping the tree lock, just like in cn_comp_spill().